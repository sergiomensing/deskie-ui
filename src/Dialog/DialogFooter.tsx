import styled from 'styled-components';
import { Button } from '../Button';
import type { DialogProps } from './Dialog';

export type DialogFooterProps = {
  actions: DialogProps['actions'];
};

const StyledDialogFooter = styled.div`
  padding: 1rem 1.5rem 1.5rem;
  display: flex;
  margin: 0 -0.5rem;
  position: sticky;
  bottom: 0;
  background: ${(props) => props.theme.colors.background};

  .button {
    margin: 0 0.5rem;
    width: 100%;
  }
`;

function DialogFooter({ actions }: DialogFooterProps) {
  return (
    <StyledDialogFooter className="dialog-footer">
      {actions?.map((action, index) => (
        <Button key={index} appearance={action.appearance || 'default'} onClick={() => action.onClick()}>
          {action.label}
        </Button>
      ))}
    </StyledDialogFooter>
  );
}

export default DialogFooter;
