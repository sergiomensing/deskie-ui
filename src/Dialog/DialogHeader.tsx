import styled from 'styled-components';
import { Text } from '../Text';
import { Button } from '../Button';
import { Close } from '@styled-icons/material-outlined';

export type DialogHeaderProps = {
  title?: string;
  onClose: () => void;
};

const StyledDialogHeader = styled.div`
  padding: 1rem 1rem 1rem 1.5rem;
  display: flex;
  align-items: center;
  position: sticky;
  top: 0;
  background: ${(props) => props.theme.colors.background};

  .button {
    margin-left: auto;
  }
`;

function DialogHeader({ title, onClose }: DialogHeaderProps) {
  return (
    <StyledDialogHeader className="dialog-header">
      <Text className="dialog-title" h2>
        {title}
      </Text>
      <Button size="small" appearance="secondary" icon={<Close />} onClick={onClose} />
    </StyledDialogHeader>
  );
}

export default DialogHeader;
