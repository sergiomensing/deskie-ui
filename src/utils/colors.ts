export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
