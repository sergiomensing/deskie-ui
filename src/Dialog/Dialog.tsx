import { PropsWithChildren, useEffect, useRef } from 'react';
import type { ButtonAppearance } from '../Button/Button';
import DialogHeader from './DialogHeader';
import DialogFooter from './DialogFooter';
import styled from 'styled-components';
import { useSpring, animated } from 'react-spring';
import { useGesture } from 'react-use-gesture';

export type DialogProps = {
  title?: string;
  className?: string;
  open?: boolean;
  onClose?: () => void;
  showHeader?: boolean;
  width?: number;
  disablePadding?: boolean;
  actions?: {
    appearance?: ButtonAppearance;
    label: string;
    onClick: () => void;
  }[];
};

const StyledDialog = styled.div<{ open: boolean; disablePadding: boolean; width: string }>`
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  pointer-events: ${(props) => (props.open ? 'initial' : 'none')};
  display: flex;
  justify-content: center;
  padding: 5vh 0;
  z-index: 999;

  @media screen and (max-width: ${(props) => props.width}) {
    padding: 10vh 0;
  }

  .dialog-underlay {
    position: fixed;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background: ${(props) => `${props.theme.colors.foreground}B3`};
    cursor: pointer;
    z-index: 1;
    will-change: opacity;
  }

  .dialog-wrapper {
    position: relative;
    touch-action: none;
    position: relative;
    z-index: 2;
    margin-top: auto;
    margin-bottom: auto;
    min-height: 10rem;
    width: min(100%, ${(props) => props.width});

    @media screen and (max-width: ${(props) => props.width}) {
      margin-bottom: 0;
    }
  }

  .dialog {
    height: 100%;
    width: 100%;
    background: ${(props) => props.theme.colors.background};
    box-shadow: ${(props) => props.theme.elevation[2]};
    border-radius: 1rem;
    -webkit-overflow-scrolling: touch;
    touch-action: pan-y;
    overflow-y: auto;
    overflow-x: hidden;
    will-change: transform opacity;

    @media screen and (max-width: ${(props) => props.width}) {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }

  .dialog-content {
    padding: ${(props) => (props.disablePadding ? '0' : '0.5rem 1.5rem')};
  }
`;

function Dialog({
  showHeader = true,
  disablePadding = false,
  open = false,
  width = 560,
  title,
  actions,
  children,
  className = ``,
  onClose,
}: PropsWithChildren<DialogProps>) {
  const dialogRef = useRef<HTMLDivElement | null>(null);

  const maxHeight = window.innerHeight * 0.9;
  const drawer = window.innerWidth <= width;

  function handleClose() {
    if (onClose) {
      onClose();
    }
  }

  const closedY = drawer ? maxHeight : 32;
  const openY = 0;

  const [{ y }, set] = useSpring<{ y: number }>(() => ({ y: closedY }));

  useEffect(() => {
    set({ y: open ? openY : closedY, immediate: false });
  }, [open]);

  const bind = useGesture(
    {
      // onScroll: ({ vxvy: [, vy], last, event, movement: [, my] }) => {
      // if (my < 0 && y.getValue() !== openY && !last) {
      //   const node = event.currentTarget as Element;
      //   node.scrollTop = 0;
      // }
      // if (last && vy < -5) {
      // Do something with the projected endpoint
      // handleClose();
      // }
      // },
      onDrag: ({ delta: [, dy], event, last, vxvy: [, vy], memo }) => {
        let _memo = (memo as number) || 0;
        const node = event.currentTarget as Element;

        if (node && (node.scrollTop <= 0 || y.getValue() !== openY)) {
          _memo = _memo + dy;
          if (last) {
            if (Math.abs(_memo) > closedY / 3 || vy > 2) {
              handleClose();
            } else {
              set({ y: openY, immediate: false });
            }
          } else {
            set({ y: Math.max(openY, _memo), immediate: true });
          }
        }

        return _memo;
      },
    },
    { drag: { useTouch: true, filterTaps: true } },
  );

  return (
    <StyledDialog
      width={`${width}px`}
      disablePadding={disablePadding}
      open={open}
      className={`dialog-container ${className}`}
    >
      <animated.div
        className="dialog-underlay"
        onClick={() => handleClose()}
        style={{ opacity: y.interpolate({ range: [openY, closedY], output: [1, 0] }) }}
      />
      <animated.div
        className="dialog-wrapper"
        style={{
          opacity: drawer ? 1 : y.interpolate({ range: [openY, closedY], output: [1, 0] }),
          transform: y.interpolate((y) => `translateY(${y}px)`),
        }}
      >
        <div ref={dialogRef} {...bind()} className="dialog" style={{ maxHeight: maxHeight }}>
          {showHeader && <DialogHeader title={title} onClose={handleClose} />}
          <div className="dialog-content">{children}</div>
          {actions && <DialogFooter actions={actions} />}
        </div>
      </animated.div>
    </StyledDialog>
  );
}

export default Dialog;
