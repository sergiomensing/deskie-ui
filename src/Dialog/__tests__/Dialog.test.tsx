import React from 'react';
import Dialog from '../Dialog';
import { fireEvent, render } from '../../../tests/utils';

describe('Dialog', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Dialog open={true} title="test-value-title" actions={[{ label: 'save', onClick: () => console.log('save') }]}>
        test-value
      </Dialog>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should close on click backdrop click', () => {
    const closeHandler = jest.fn();
    const { container } = render(
      <Dialog open={true} onClose={closeHandler} title="test-value-title">
        test-value
      </Dialog>,
    );

    fireEvent.click(container.querySelector('.dialog-underlay'));

    expect(closeHandler).toHaveBeenCalledTimes(1);
  });

  it('should show and trigger actions', () => {
    const actionHandler = jest.fn();

    const { getByText } = render(
      <Dialog open={true} showHeader={false} actions={[{ label: 'test-value-action', onClick: actionHandler }]}>
        test-value
      </Dialog>,
    );

    fireEvent.click(getByText('test-value-action'));
    expect(actionHandler).toHaveBeenCalledTimes(1);
  });

  it('should not render footer if no actions are defined', () => {
    const { container } = render(<Dialog open={true}>test-value</Dialog>);

    expect(container.querySelector('.dialog-footer')).not.toBeInTheDocument();
  });

  it('should not render header if showHeader equals false', () => {
    const { container } = render(
      <Dialog open={true} showHeader={false}>
        test-value
      </Dialog>,
    );

    expect(container.querySelector('.dialog-header')).not.toBeInTheDocument();
  });
});
