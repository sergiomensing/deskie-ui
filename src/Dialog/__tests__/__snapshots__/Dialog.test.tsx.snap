// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should render correctly 1`] = `
.c2 {
  font-size: 1.125rem;
  font-weight: 500;
  font-family: 'Poppins',sans-serif;
  color: #11142D;
}

.c3 {
  border: 1px solid transparent;
  font: inherit;
  height: 2.5rem;
  border-radius: 0.75rem;
  padding: 0 0.75rem;
  width: 2.5rem;
  background-color: transparent;
  border-color: transparent;
  color: #11142D;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c3:hover {
  background-color: #F7F7FD;
  border-color: #F7F7FD;
}

.c3 .button-icon {
  width: 1.25rem;
}

.c3 .button-icon svg {
  width: 100%;
  height: auto;
}

.c6 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  padding: 0 1.25rem;
  background-color: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c6:hover {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c6 .button-icon {
  width: 1.25rem;
}

.c6 .button-icon svg {
  width: 100%;
  height: auto;
}

.c6 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c6 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  padding: 1rem 1rem 1rem 1.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  background: #FFFFFF;
}

.c1 .button {
  margin-left: auto;
}

.c5 {
  padding: 1rem 1.5rem 1.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 -0.5rem;
  position: -webkit-sticky;
  position: sticky;
  bottom: 0;
  background: #FFFFFF;
}

.c5 .button {
  margin: 0 0.5rem;
  width: 100%;
}

.c0 {
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  pointer-events: initial;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 5vh 0;
  z-index: 999;
}

.c0 .dialog-underlay {
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background: #11142DB3;
  cursor: pointer;
  z-index: 1;
  will-change: opacity;
}

.c0 .dialog-wrapper {
  position: relative;
  touch-action: none;
  position: relative;
  z-index: 2;
  margin-top: auto;
  margin-bottom: auto;
  min-height: 10rem;
  width: min(100%,560px);
}

.c0 .dialog {
  height: 100%;
  width: 100%;
  background: #FFFFFF;
  box-shadow: 0px 1px 2px rgba(0,0,0,0.02),0px 3px 7px rgba(0,0,0,0.03),0px 12px 31px rgba(0,0,0,0.05);
  border-radius: 1rem;
  -webkit-overflow-scrolling: touch;
  touch-action: pan-y;
  overflow-y: auto;
  overflow-x: hidden;
  will-change: transform opacity;
}

.c0 .dialog-content {
  padding: 0.5rem 1.5rem;
}

@media screen and (max-width:560px) {
  .c0 {
    padding: 10vh 0;
  }
}

@media screen and (max-width:560px) {
  .c0 .dialog-wrapper {
    margin-bottom: 0;
  }
}

@media screen and (max-width:560px) {
  .c0 .dialog {
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
  }
}

<div>
  <div
    class="c0 dialog-container "
    open=""
    width="560px"
  >
    <div
      class="dialog-underlay"
      style="opacity: 0;"
    />
    <div
      class="dialog-wrapper"
      style="opacity: 0; transform: translateY(32px);"
    >
      <div
        class="dialog"
        style="max-height: 691.2px;"
      >
        <div
          class="c1 dialog-header"
        >
          <h2
            class="c2 dialog-title"
          >
            test-value-title
          </h2>
          <button
            class="c3 button button-iconOnly "
          >
            <span
              class="button-icon"
            >
              <svg
                aria-hidden="true"
                class="c4"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="dialog-content"
        >
          test-value
        </div>
        <div
          class="c5 dialog-footer"
        >
          <button
            class="c6 button  "
          >
            save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
